File: activemq-dotnet/src/test/csharp/NMS/JMSTestSupport.cs
/*
 * Copyright 2006 The Apache Software Foundation or its licensors, as
 * applicable.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using NMS;
using NUnit.Framework;
using System;

/// <summary>
/// useful base class for test cases
/// </summary>

namespace NMS
{
	[ TestFixture ]
    public abstract class JMSTestSupport
    {
        
		private IConnectionFactory factory;
        private IConnection connection;
		private ISession session;
		private IDestination destination;
		
		protected int receiveTimeout = 1000;
		protected string clientId;
		protected DestinationType destinationType = DestinationType.Queue;
		protected AcknowledgementMode acknowledgementMode = AcknowledgementMode.ClientAcknowledge;
		
		[SetUp]
        virtual public void SetUp()
        {
        }
		
        [TearDown]
        virtual public void TearDown()
        {
			Disconnect();
        }
		
		// Properties
        public bool Connected
        {
            get { return connection!=null; }
            set { if( value ) Connect(); else Disconnect(); }
        }

        public IConnectionFactory Factory
        {
            get {
				if( factory == null ) {
					factory = CreateConnectionFactory();
					Assert.IsNotNull(factory, "no factory created");
				}
				return factory;
			}
            set { this.factory = value; }
        }
		
        public IConnection Connection
        {
            get {
				if( connection == null ) {
					Connect();
				}
				return connection;
			}
            set { this.connection = value; }
        }
		
        public ISession Session
        {
            get {
				if( session == null ) {
					session = Connection.CreateSession(acknowledgementMode);
					Assert.IsNotNull(connection != null, "no session created");
				}
				return session;
			}
            set { this.session = value; }
        }
		
		virtual protected void Connect()
        {
			Console.WriteLine("Connectting...");
			connection = CreateConnection();
			Assert.IsNotNull(connection, "no connection created");
			connection.Start();
			Console.WriteLine("Connected.");
			Assert.IsNotNull(connection, "no connection created");
        }
		        
        virtual protected void Disconnect()
        {
            if (connection != null)
            {
				Console.WriteLine("Disconnecting...");
                connection.Dispose();
                connection = null;
				session=null;
				Console.WriteLine("Disconnected.");
            }
        }
        
        virtual protected void Reconnect()
        {
            Disconnect();
            Connect();
        }
		
		protected virtual void Drain()
		{
            using (ISession session = Connection.CreateSession())
            {
				// Tries to consume any messages on the Destination
				IMessageConsumer consumer = session.CreateConsumer(Destination);
				
				// Should only need to wait for first message to arrive due to the way
				// prefetching works.
				IMessage msg = consumer.Receive(TimeSpan.FromMilliseconds(receiveTimeout));
				while (msg != null)
				{
					msg = consumer.ReceiveNoWait();
				}
			}
		}
		
        public virtual void SendAndSyncReceive()
        {
            using (ISession session = Connection.CreateSession())
            {
				
				IMessageConsumer consumer = session.CreateConsumer(Destination);
				IMessageProducer producer = session.CreateProducer(Destination);
				
				IMessage request = CreateMessage();
				producer.Send(request);
				
				IMessage message = consumer.Receive(TimeSpan.FromMilliseconds(receiveTimeout));
                Assert.IsNotNull(message, "No message returned!");
                AssertValidMessage(message);
            }
        }
		
		protected virtual IConnectionFactory CreateConnectionFactory()
		{
			return new ActiveMQ.ConnectionFactory(new Uri("tcp://localhost:61616"));
		}
		
		protected virtual IConnection CreateConnection()
		{
			IConnection connection =  Factory.CreateConnection();
			if( clientId!=null ) {
				connection.ClientId = clientId;
			}
			return connection;
		}
		
		protected virtual IMessageProducer CreateProducer()
		{
			IMessageProducer producer = Session.CreateProducer(destination);
			return producer;
		}
		
		protected virtual IMessageConsumer CreateConsumer()
		{
			IMessageConsumer consumer = Session.CreateConsumer(destination);
			return consumer;
		}
        
        protected virtual IDestination CreateDestination()
        {
			if( destinationType == DestinationType.Queue ) {
				return Session.GetQueue(CreateDestinationName());
			} else if( destinationType == DestinationType.Topic ) {
				return Session.GetTopic(CreateDestinationName());
			} else if( destinationType == DestinationType.TemporaryQueue ) {
				return Session.CreateTemporaryQueue();
			} else if( destinationType == DestinationType.TemporaryTopic ) {
				return Session.CreateTemporaryTopic();
			} else {
				throw new Exception("Unknown destination type: "+destinationType);
			}
        }
		
        protected virtual string CreateDestinationName()
        {
            return "Test.DotNet." + GetType().Name;
        }
        
        protected virtual IMessage CreateMessage()
        {
            return Session.CreateMessage();
        }
        
        protected virtual  void AssertValidMessage(IMessage message)
        {
            Assert.IsNotNull(message, "Null Message!");
        }
		
		
        public IDestination Destination
        {
            get {
				if (destination == null)
				{
					destination = CreateDestination();
					Assert.IsNotNull(destination, "No destination available!");
					Console.WriteLine("Using destination: " + destination);
				}
				return destination;
			}
            set {
				destination = value;
            }
        }
		
    }
}

File: activemq-core/src/test/java/org/apache/activemq/JmsTestSupport.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.activemq;

import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

import javax.jms.Connection;
import javax.jms.ConnectionFactory;
import javax.jms.Destination;
import javax.jms.JMSException;
import javax.jms.MessageConsumer;
import javax.jms.MessageProducer;
import javax.jms.Session;

import org.apache.activemq.broker.BrokerFactory;
import org.apache.activemq.broker.BrokerService;
import org.apache.activemq.command.ActiveMQDestination;

/**
 * Test cases used to test the JMS message consumer.
 * 
 * 
 */
public class JmsTestSupport extends CombinationTestSupport {

    static final private AtomicLong TEST_COUNTER = new AtomicLong();
    public String userName;
    public String password;
    public String messageTextPrefix = "";

    protected ConnectionFactory factory;
    protected ActiveMQConnection connection;
    protected BrokerService broker;

    protected List<Connection> connections = Collections.synchronizedList(new ArrayList<Connection>());

    // /////////////////////////////////////////////////////////////////
    //
    // Test support methods.
    //
    // /////////////////////////////////////////////////////////////////
    protected ActiveMQDestination createDestination(Session session, byte type) throws JMSException {
        String testMethod = getName();
        if( testMethod.indexOf(" ")>0 ) {
            testMethod = testMethod.substring(0, testMethod.indexOf(" "));
        }
        String name = "TEST." + getClass().getName() + "." +testMethod+"."+TEST_COUNTER.getAndIncrement();
        switch (type) {
        case ActiveMQDestination.QUEUE_TYPE:
            return (ActiveMQDestination)session.createQueue(name);
        case ActiveMQDestination.TOPIC_TYPE:
            return (ActiveMQDestination)session.createTopic(name);
        case ActiveMQDestination.TEMP_QUEUE_TYPE:
            return (ActiveMQDestination)session.createTemporaryQueue();
        case ActiveMQDestination.TEMP_TOPIC_TYPE:
            return (ActiveMQDestination)session.createTemporaryTopic();
        default:
            throw new IllegalArgumentException("type: " + type);
        }
    }

    protected void sendMessages(Destination destination, int count) throws Exception {
        ConnectionFactory factory = createConnectionFactory();
        Connection connection = factory.createConnection();
        connection.start();
        sendMessages(connection, destination, count);
        connection.close();
    }

    protected void sendMessages(Connection connection, Destination destination, int count) throws JMSException {
        Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
        sendMessages(session, destination, count);
        session.close();
    }

    protected void sendMessages(Session session, Destination destination, int count) throws JMSException {
        MessageProducer producer = session.createProducer(destination);
        for (int i = 0; i < count; i++) {
            producer.send(session.createTextMessage(messageTextPrefix  + i));
        }
        producer.close();
    }

    protected ConnectionFactory createConnectionFactory() throws Exception {
        return new ActiveMQConnectionFactory("vm://localhost");
    }

    protected BrokerService createBroker() throws Exception {
        return BrokerFactory.createBroker(new URI("broker://()/localhost?persistent=false"));
    }

    protected void setUp() throws Exception {
        super.setUp();

        if (System.getProperty("basedir") == null) {
            File file = new File(".");
            System.setProperty("basedir", file.getAbsolutePath());
        }

        broker = createBroker();
        broker.start();
        factory = createConnectionFactory();
        connection = (ActiveMQConnection)factory.createConnection(userName, password);
        connections.add(connection);
    }

    protected void tearDown() throws Exception {
        for (Iterator iter = connections.iterator(); iter.hasNext();) {
            Connection conn = (Connection)iter.next();
            try {
                conn.close();
            } catch (Throwable e) {
            }
            iter.remove();
        }
        broker.stop();
        super.tearDown();
    }

    protected void safeClose(Connection c) {
        try {
            c.close();
        } catch (Throwable e) {
        }
    }

    protected void safeClose(Session s) {
        try {
            s.close();
        } catch (Throwable e) {
        }
    }

    protected void safeClose(MessageConsumer c) {
        try {
            c.close();
        } catch (Throwable e) {
        }
    }

    protected void safeClose(MessageProducer p) {
        try {
            p.close();
        } catch (Throwable e) {
        }
    }

    protected void profilerPause(String prompt) throws IOException {
        if (System.getProperty("profiler") != null) {
            pause(prompt);
        }
    }

    protected void pause(String prompt) throws IOException {
        System.out.println();
        System.out.println(prompt + "> Press enter to continue: ");
        while (System.in.read() != '\n') {
        }
    }

}
